proposals/infra/INF-0004-validator-hashing.md

```diff
@@ -123,42 +124,140 @@ should be that all shaders are validated and hashed using what is today the
 
-### Changes for D3D
+### D3D Runtime Behavior
 
-A future AgilitySDK will adopt runtime changes to how and when the shader hash
-is validated and add support for DXIL validation in the debug layer and runtime.
+Starting with AgilitySDK version TBD, the runtime adopts new behavior for how 
+and when the shader hash is validated, including support for bytecode validation 
+in the debug layer and runtime.
 
-For shaders with the `PREVIEW` hash, the shader is allowed to execute only if
-developer mode and the experimental feature D3D12ExperimentalShaderModels is
-enabled, otherwise the runtime will produce an error.
+For shaders with the `PREVIEW_BYPASS` hash, the shader is allowed to execute 
+only if developer mode and the experimental feature 
+`D3D12ExperimentalShaderModels` is enabled, otherwise the runtime will produce
+an error.
 
-For shaders with the `BYPASS` hash, the new behavior will allow shaders to run
-without validating the hash regardless of whether the machine is in developer
-mode as long as the shader targets a supported non-experimental shader model.
+For shaders with the `BYPASS` hash, shaders can run without validating the hash 
+regardless of whether the machine is in developer mode as long as the shader 
+targets a supported non-experimental shader model.
 
-All other values of the hash will be assumed to be real hash values.
+All other values of the hash are assumed to be real hash values.
 
 For shaders that contain a real hash value, the behavior is unchanged. The hash
-will be verified and the shader will execute as it does with existing versions
-of the runtime.
+is verified and the shader executes as it does with existing versions of the 
+runtime.
 
 > Note: an important behavior change here is that shaders with zero'd hash data
 > will always be treated as invalid by the runtime and rejected.
+> 
+> An exception is for a period of time already existing preview shaders that used 
+> zero'd hash data will be treated like having the `PREVIEW_BYPASS` hash in that
+> preview runtimes will allow them to be used, at least until those preview shaders
+> and runtimes are no longer supported.  In contrast to `PREVIEW_BYPASS` shaders,
+> these legacy zero'd has shaders will not run have the validator invoked on them
+> by the runtime/debug layer since that wasn't expected of them.
+
+#### D3D Runtime Validation Control
+
+Apps can configure how the runtime validates bytecode passed to it. 
+
+The validation referred to here is implementation exposed by `dxil.dll` 
+or `dxcompiler.dll`, and also embedded in `d3d12sdklayers.dll`.  
+This is the same validation used when the compiler endorses bytecode at 
+compile time by applying a hash.  So it would typically be 
+redundant to validate again, but it can be useful to validate shaders that have 
+the `BYPASS` or `PREVIEW_BYPASS` hash, or have the option to force validation. 
+
+If there is bytecode to validate, the runtime and/or debug layer attempt to 
```

Comment:
> the runtime and/or debug layer

Just to clarify here (because I was under the impression that only the debug layer was ever going to do validation).  Is the proposal here that somethings the runtime will do validation (by calling into some external DLL) but at other times the debug layer will do validation (using a version statically linked with it)?

Or something else?

Reply:
The flags let the user decide what they want.   All the details are there, but brief summary:
- by default only debug layer validates only BYPASS/BYPASS_PREVIEW shaders, printing messages and not failing. The debug layer doesn't alter pass/fail behavior by design (to the extent possible)
- the app can choose to have the runtime validate as well, with or without debug layer.  When the runtime is validating, it will fail (e.g. state object creation fail) when it sees an error.  So an app could choose this mode if it wants just an extra layer of checking on any shader it gets in the field in retail (e.g. without debug layer)
- the app can choose for the validation to apply to all shaders, even hashed ones, if it wants.   Maybe they want to try a new validator.
- The app can choose where the validator implementation comes from, incl from debug layer (when enabled) or dxil.dll/dxcompiler.dll whichever is present.  The search order can be tweaked.
- The app can choose to skip validation if an implementation isn't available, or have validation fail if an implementation isn't available
- If the debug layer and runtime are both doing validation (runtime for pass/fail, debug layer for printing messages), they will always select the same validator implementation so they are consistent.   This pattern is similar to how a huge amount of existing runtime+debug layer validation works - they are guaranteed to be consistent with each other.

Reply:
I can see that the proposal adds a huge amount of flexibility.  I'm unsure, though, why we're planning on doing any of that and why it is necessary?

Reply:
This is defining what was originally proposed, which is that there would be controls on when to validate.

If we're sanctioning the workflow of bypassing the validator during author time, the least we should do is let apps validate at runtime.  There isn't any extra burden on HLSL team here since this is validation code the app can also run themselves manually.  But I think it is of great utility to make it straightforward to perform the validation in all the paths and engine ends up feeding shaders to D3D by simply having the runtime do it when it sees a shader (vs the app having to locate all the paths where they might be sourcing shaders from.

Reply:
Per offline discussion, removed runtime validation support.